src/search/plugins/plugin.h

```diff
@@ -1,103 +1,245 @@
 #include "raw_registry.h"
-#include "type_namer.h"
 
 #include "../utils/strings.h"
 #include "../utils/system.h"
 
-#include <algorithm>
-#include <functional>
-#include <memory>
 #include <string>
 #include <typeindex>
-#include <typeinfo>
-#include <utility>
+#include <type_traits>
 #include <vector>
 
 namespace plugins {
-class Registry;
+class Feature {
+    const Type &type;
+    std::string key;
+    std::string title;
+    std::string synopsis;
+    std::string subcategory;
+    std::vector<ArgumentInfo> arguments;
+    std::vector<PropertyInfo> properties;
+    std::vector<LanguageSupportInfo> language_support;
+    std::vector<NoteInfo> notes;
+public:
+    Feature(const Type &type, const std::string &key);
+    virtual ~Feature() = default;
+    Feature(const Feature &) = delete;
 
-/*
-  The following function is not meant for users, but only for the
-  plugin implementation. We only declare it here because the template
-  implementations need it.
-*/
-extern void register_plugin_type_plugin(
-    const std::type_info &type,
-    const std::string &type_name,
-    const std::string &documentation);
+    virtual Any construct(const Options &opts, const ConstructContext &context) const = 0;
+
+    /* Add option with default value. Use def_val=ArgumentInfo::NO_DEFAULT for
+       optional parameters without default values. */
+    template<typename T>
+    void add_option(
+        const std::string &key,
+        const std::string &help = "",
+        const std::string &default_value = "",
+        const Bounds &bounds = Bounds::unlimited(),
+        bool lazy_construction = false);
+
+    template<typename T>
+    void add_list_option(
+        const std::string &key,
+        const std::string &help = "",
+        const std::string &default_value = "",
+        bool lazy_construction = false);
+
+    void document_subcategory(const std::string &subcategory);
+    void document_title(const std::string &title);
+    void document_synopsis(const std::string &note);
+    void document_property(
+        const std::string &property, const std::string &note);
+    void document_language_support(
+        const std::string &feature, const std::string &note);
+    void document_note(
+        const std::string &title, const std::string &note, bool long_text = false);
+
+    const Type &get_type() const;
+    std::string get_key() const;
+    std::string get_title() const;
+    std::string get_synopsis() const;
+    std::string get_subcategory() const;
+    const std::vector<ArgumentInfo> &get_arguments() const;
+    const std::vector<PropertyInfo> &get_properties() const;
+    const std::vector<LanguageSupportInfo> &get_language_support() const;
+    const std::vector<NoteInfo> &get_notes() const;
+};
+
+
+template<typename Constructed>
+class FeatureWithDefault : public Feature {
+protected:
+    using Feature::Feature;
+    virtual std::shared_ptr<Constructed> create_component(const Options &options, const ConstructContext &) const {
```

Comment:
The `&` belongs to the type (`const ConstructContext &`), so we are passing a const reference to a construct context here. We left out the name, because the context is not used here and its name is not relevant. The type already explains everything the variable name could explain. We could mention the name as a comment (`const ConstructContext & /* context */`)but I don't think it is necessary here. Searching for `&)` and `& /*` in our code, it seems we did both so far, but leaving out the variable name is more common. I'd leave this as is.